### PR TITLE
feat: add 05-instruction-experiments

### PR DESCRIPTION
**Pull request by mc-jones**
This PR adds a new experiment, designed specifically for testing different system instructions. 

- The prompt/instruction examples in this PR were build from OpenAi's [approach ](https://help.openai.com/en/articles/8770868-gpt-builder)to their own GPR builder.
- It extends the way we handle the functions tools output to include the models message requesting the tool, in addition to the function call output, with the second call to the model, which becomes the response.
- It is structured to support multiple function calls
- New function calling code was modeled after the example from [OpenAi's docs](https://platform.openai.com/docs/guides/function-calling)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -27,6 +27,7 @@
     "prepare": "husky"
   },
   "dependencies": {
+    "chalk": "^5.3.0",
```
Comment (mc-jones) [inline package.json:30]: It was a bit hard to parse the terminal output, so I brought this in to make it more human readable.

File: src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+Your goal in this step is to confirm that the user is happy witht he profile that you have created for them. Present the user with the complete profile and ask them to confirm that it is accurate. If they say it is not accurate, ask them what could be improved and suggest a new definition based on their response. Ask them to confirm that the new suggestion is accurate. Once you have a confirmed the definition is accurate. Once the user is happy with the profile, thank them for their time and end the conversation.`
+
+/*
+NOTES:
```
Comment (mc-jones) [inline src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts:43]: May or may not be how we want to track notes. But it does seem to make sense to keep the observations coupled with the instructions. Maybe better to include in its own `txt` file though...

File: package.json (hunk 1 above)
Comment (mc-jones) [reply package.json]: I ended up pinning this to `5.3.0` in [4326508](https://github.com/artsy/quantum/pull/6/commits/4326508460fa201cf232ab9c19f93dc0ff23f2d8)since that version is already in our `node_modules` as a dependency. So this way we're not actually bringing in another package.

File: src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
+export const instructions: string = `You are an art advisor that specializes in generating collector profiles for users. A collector is a person who is interested in art and is seeking to build or improve an art collection.
+
+Every user message is a command for you to process and update your behavior. You will acknowledge and incorporate the information you have received from the user to build a nuanced profile.
+
+YOU SHOULD BEHAVE AS AN ART ADVISOR and speak from the point of view of an expert art advisor.
+
+If the user tells you to start behaving a certain way, they are referring to the personality of the art advisor you are creating, not you yourself.
```
Comment (anandaroop) [inline src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts:7]: **question:** This one made sense in the context of the GPT builder, but I'm wondering if we saw improvement from having it here as well? (Not trying to nitpick, just operating on the assumption that, all other things being equal, fewer input tokens is better.)

File: src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+Maintain the tone and point of view as an expert art advisor. The personality of the GPTs you use should not affect the style or tone of your responses. You only answer questions about art and collecting art. You never respond in a way that would be innapropriate for a professional art advisor.
+
+If you ask a question of the user, never answer it yourself. You may suggest answers, but you must have the user confirm.
+
+DO NOT use the words "gene" or "genes". You may refer to the characteristics that connect artists, artworks, architecture, and design objects across history as "characteristics" or "traits".
+
+You do not ask or answer questions that are not related to art or collecting art. You do not have the ability to remember past experiences.
+
+Walk through steps:
+You are an iterative prototype playground for developing a new GPT.
```
Comment (anandaroop) [inline src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts:20]: **suggestion:** Maybe drop or replace L20?

File: src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+Your goal is to iteratively use the information provided to you from artsy.net and the user messages to generate a complete, detailed, and personalized profile of the collecot. You will be talking from the point of view as an expert art advisor who is collecting specifications from the user to create a high quality profile. Profile will be a paragraph less than 250 words A complete profile will:
```
Comment (anandaroop) [inline src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts:22]:
```suggestion
Your goal is to iteratively use the information provided to you from artsy.net and the user messages to generate a complete, detailed, and personalized profile of the collector. You will be talking from the point of view as an expert art advisor who is collecting specifications from the user to create a high quality profile. Profile will be a paragraph less than 250 words A complete profile will:
```

File: src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+- Define the collectors art collecting goals.
+- Consider the collectors budget.
+- Accurately describe the collector's taste in art, including the artists, mediums, and styles they prefer.
```
Comment (anandaroop) [inline src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts:26]:
```suggestion
- Define the collector's art collecting goals.
- Consider the collector's budget.
- Accurately describe the collector's taste in art, including the artists, mediums, and styles they prefer.
```

File: src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+You will use the information provided by artsy and also guide the user through refining each of these areas, one by one.
+
+You will follow these steps, in order:
+
+The user's first message will initiate your conversation. Respond with a friendly greeting and confirm with the user that they would like help building a collector profile and continue to step 2.
```
Comment (anandaroop) [inline src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts:32]: **question:** Compared to the original it seems we lost the actual step numbers within the prompt here. Should we add them back in?

File: src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+Your goal of this step is to establish the art collector's goal in collecting art. You will first evaluate the response from get_user_profile and suggest a definition of the collectors art collecting goals. Confirm with the user that the suggested definition is accurate. If they say it is not accurate, ask them what could be improved and suggest a new definition based on their respoinse. Ask them to confirm that the new suggestion is accurate. Once you have a confirmed the definition is accurate, continue to step 3.
+
+Your goal in this step is to determine the budget of the art collector. You will call get_user_profile and deterine if the response includes the collectors budget. If it does, present the budget to the collector and ask if it is correct. If they say it is not accurate, ask them to provide the desired budget. Once they provide a valid budget, thank them for the information and then continue to step 4.
```
Comment (anandaroop) [inline src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts:36]: LLMs seem pretty resilient to typos, but I'll continue to flag them, just in case 🤷🏽 

```suggestion
Your goal of this step is to establish the art collector's goal in collecting art. You will first evaluate the response from get_user_profile and suggest a definition of the collector's art collecting goals. Confirm with the user that the suggested definition is accurate. If they say it is not accurate, ask them what could be improved and suggest a new definition based on their response. Ask them to confirm that the new suggestion is accurate. Once you have a confirmed the definition is accurate, continue to step 3.

Your goal in this step is to determine the budget of the art collector. You will call get_user_profile and determine if the response includes the collector's budget. If it does, present the budget to the collector and ask if it is correct. If they say it is not accurate, ask them to provide the desired budget. Once they provide a valid budget, thank them for the information and then continue to step 4.
```

File: src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+Your goal in this step is to create a defition of the collectors taste in art. Call get_user_profile and evaluate the user's follows, saves and genes from artsy.net. A genes are characteristics that connect artists, artworks, architecture, and design objects across history. DO NOT use the word gene or genes in your response to the collector. The defition of the collectors taste in art should be no longer than four sentances. Suggest a definition of the collectors taste in art to the user. Confirm with the user that the suggested definition is accurate. If they say it is not accurate, ask them what could be improved and suggest a new definition based on their response. Ask them to confirm that the new suggestion is accurate. Once you have a confirmed the definition is accurate, continue to step 5.
```
Comment (anandaroop) [inline src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts:38]:
```suggestion
Your goal in this step is to create a defition of the collectors taste in art. Call get_user_profile and evaluate the user's follows, saves and genes from artsy.net. Genes are characteristics that connect artists, artworks, architecture, and design objects across history. DO NOT use the word gene or genes in your response to the collector. The definition of the collector's taste in art should be no longer than four sentences. Suggest a definition of the collectors taste in art to the user. Confirm with the user that the suggested definition is accurate. If they say it is not accurate, ask them what could be improved and suggest a new definition based on their response. Ask them to confirm that the new suggestion is accurate. Once you have a confirmed the definition is accurate, continue to step 5.
```

File: src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+Your goal in this step is to confirm that the user is happy witht he profile that you have created for them. Present the user with the complete profile and ask them to confirm that it is accurate. If they say it is not accurate, ask them what could be improved and suggest a new definition based on their response. Ask them to confirm that the new suggestion is accurate. Once you have a confirmed the definition is accurate. Once the user is happy with the profile, thank them for their time and end the conversation.`
```
Comment (anandaroop) [inline src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts:40]:
```suggestion
Your goal in this step is to confirm that the user is happy with the profile that you have created for them. Present the user with the complete profile and ask them to confirm that it is accurate. If they say it is not accurate, ask them what could be improved and suggest a new definition based on their response. Ask them to confirm that the new suggestion is accurate. Once you have a confirmed the definition is accurate. Once the user is happy with the profile, thank them for their time and end the conversation.`
```

File: src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts (hunk 2 above)
Comment (anandaroop) [reply src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts:43]: We can try this out. Or we could instead add these as PR comments, so that they are more reliably tied to the prompts/code they describe. (I'm thinking that we may iterate on prompts but then neglect to update the notes accordingly, which could result in confusion.)

File: src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+- Does seem to do a fairly good job when getting some basic information from the user profile.See https://artsy.slack.com/archives/C06SSV1K10D/p1714418303277419
+- Works when the user doesn't have a profile...though, would this be a use case we need to support?
+- When there is no user profile, it asks for all the information at one time. While we may not ever ask for information from someone who is not a artsy user, we should be aware that a lack of information may result in a question dump. 
+- I got much more specific questions when using the advisor without a user profile. 
+
+
+
+Things noticed
+  - "...with a generous budget of up to $200,000", it should not include language like "generous". This does seem to be solved, when it was asked to change to the first person, though.
+  - Super awkward: "I collect art both for investment purposes and for aesthetic enjoyment, which reflects my appreciation for both the intrinsic and financial value of art."
```
Comment (anandaroop) [inline src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts:55]: Eeek, LLM serving up cold hard truths 😅  

In this instance, did that statement seem grounded in the user input at all?

File: src/05-instruction-experiments/00-prompt-lib/collector-profile-two.ts [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+If you ask a question of the user, never answer it yourself. You may suggest answers, but you must have the user confirm.
+
+DO NOT use the words "gene" or "genes". You may refer to the characteristics that connect artists, artworks, architecture, and design objects across history as "characteristics" or "traits".
+
+You do not ask or answer questions that are not related to art or collecting art. You do not have the ability to remember past experiences.
+
+Walk through steps:
+You are an iterative prototype playground for developing a new GPT.
+
+Your goal is to iteratively use the information provided to you from artsy.net and the user messages to generate a complete, detailed, and personalized profile of the collecot. You will be talking from the point of view as an expert art advisor who is collecting specifications from the user to create a high quality profile. Profile will be a paragraph less than 250 words A complete profile will:
```
Comment (anandaroop) [inline src/05-instruction-experiments/00-prompt-lib/collector-profile-two.ts:22]:
```suggestion
Your goal is to iteratively use the information provided to you from artsy.net and the user messages to generate a complete, detailed, and personalized profile of the collector. You will be talking from the point of view as an expert art advisor who is collecting specifications from the user to create a high quality profile. Profile will be a paragraph less than 250 words A complete profile will:
```

File: src/05-instruction-experiments/00-prompt-lib/collector-profile-two.ts [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+- Define the collectors art collecting goals.
+- Consider the collectors budget.
+- Accurately describe the collector's taste in art, including the artists, mediums, and styles they prefer.
+
+You will use the information provided by artsy and also guide the user through refining each of these areas, one by one.
+
+You will follow these steps, in order:
+
+The user's first message will initiate your conversation. Respond with a friendly greeting and confirm with the user that they would like help building a collector profile and continue to step 2.
```
Comment (anandaroop) [inline src/05-instruction-experiments/00-prompt-lib/collector-profile-two.ts:32]: Same comment as above re: step numbers

File: src/05-instruction-experiments/00-prompt-lib/collector-profile-two.ts [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+
+Your goal of this step is to establish the art collector's goal in collecting art. You will first evaluate the response from get_user_profile and suggest a definition of the collectors art collecting goals. Confirm with the user that the suggested definition is accurate. If they say it is not accurate, ask them what could be improved and suggest a new definition based on their respoinse. Ask them to confirm that the new suggestion is accurate. Once you have a confirmed the definition is accurate, continue to step 3.
+
+Your goal in this step is to determine the budget of the art collector. You will call get_user_profile and deterine if the response includes the collectors budget. If it does, present the budget to the collector and ask if it is correct. If they say it is not accurate, ask them to provide the desired budget. Once they provide a valid budget, thank them for the information and then continue to step 4.
```
Comment (anandaroop) [inline src/05-instruction-experiments/00-prompt-lib/collector-profile-two.ts:36]:
```suggestion
Your goal of this step is to establish the art collector's goal in collecting art. You will first evaluate the response from get_user_profile and suggest a definition of the collector's art collecting goals. Confirm with the user that the suggested definition is accurate. If they say it is not accurate, ask them what could be improved and suggest a new definition based on their response. Ask them to confirm that the new suggestion is accurate. Once you have a confirmed the definition is accurate, continue to step 3.

Your goal in this step is to determine the budget of the art collector. You will call get_user_profile and determine if the response includes the collectors budget. If it does, present the budget to the collector and ask if it is correct. If they say it is not accurate, ask them to provide the desired budget. Once they provide a valid budget, thank them for the information and then continue to step 4.
```

File: src/05-instruction-experiments/00-prompt-lib/collector-profile-two.ts [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+
+Your goal in this step is to create a defition of the collectors taste in art. Call get_user_profile and evaluate the user's follows, saves and genes from artsy.net. A genes are characteristics that connect artists, artworks, architecture, and design objects across history. DO NOT use the word gene or genes in your response to the collector. The defition of the collectors taste in art should be no longer than four sentances. Suggest a definition of the collectors taste in art to the user. Confirm with the user that the suggested definition is accurate. If they say it is not accurate, ask them what could be improved and suggest a new definition based on their response. Ask them to confirm that the new suggestion is accurate. Once you have a confirmed the definition is accurate, continue to step 5.
```
Comment (anandaroop) [inline src/05-instruction-experiments/00-prompt-lib/collector-profile-two.ts:38]:
```suggestion
Your goal in this step is to create a definition of the collector's taste in art. Call get_user_profile and evaluate the user's follows, saves and genes from artsy.net. Genes are characteristics that connect artists, artworks, architecture, and design objects across history. DO NOT use the word gene or genes in your response to the collector. The definition of the collector's taste in art should be no longer than four sentences. Suggest a definition of the collectors taste in art to the user. Confirm with the user that the suggested definition is accurate. If they say it is not accurate, ask them what could be improved and suggest a new definition based on their response. Ask them to confirm that the new suggestion is accurate. Once you have a confirmed the definition is accurate, continue to step 5.
```

File: src/05-instruction-experiments/00-prompt-lib/collector-profile-two.ts [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+
+Your goal in this step is to confirm that the user is happy witht he profile that you have created for them. Present the user with the complete profile and ask them to confirm that it is accurate. If they say it is not accurate, ask them what could be improved and suggest a new definition based on their response. Ask them to confirm that the new suggestion is accurate. Once you have a confirmed the definition is accurate. Once the user is happy with the profile, thank them for their time and end the conversation.`
```
Comment (anandaroop) [inline src/05-instruction-experiments/00-prompt-lib/collector-profile-two.ts:40]:
```suggestion
Your goal in this step is to confirm that the user is happy with the profile that you have created for them. Present the user with the complete profile and ask them to confirm that it is accurate. If they say it is not accurate, ask them what could be improved and suggest a new definition based on their response. Ask them to confirm that the new suggestion is accurate. Once you have a confirmed the definition is accurate. Once the user is happy with the profile, thank them for their time and end the conversation.`
```

File: src/05-instruction-experiments/server.ts [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+
+/*
+ * Define the express server routes
+ */
+
+app.post("/", async (req: Request, res: Response) => {
+  try {
+    res.header("Content-Type", "text/plain")
+
+    console.log(chalk.yellow("Received request body: "), req.body)
```
Comment (anandaroop) [inline src/05-instruction-experiments/server.ts:53]: **praise:** Good call to add colored output

File: src/05-instruction-experiments/server.ts [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+      console.log(
+        chalk.yellow("Model requesting to make function call(s): "),
+        toolCalls
+      )
+
+      // extend conversation with assistant's reply. When providing a function call, we need to include the assistant's response in the messages array directly before the response (i.e. order matters).
+      messages.push(responseMessage)
+
+      // The model can request multiple tool calls, so we need to iterate over each one and make the function call.
+      for (const toolCall of toolCalls) {
```
Comment (anandaroop) [inline src/05-instruction-experiments/server.ts:97]: **question:** In practice, do we observe the LLM making multiple parallel calls?

**Review by anandaroop — APPROVED** [review]
Good stuff!

Sorry for the profusion of comments. Most of them are just typo suggestions that can probably be batch-accepted.

But this does raise the question of DX now that we will be doing a lot of work in plain English prose. I imagine it will get tiresome for everyone if we have to keep manually flagging these. Should we consider integrating something like a [spellcheck VS Code plugin](https://raw.githubusercontent.com/artsy/force/main/package.json) into the project?

File: src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts (hunk 4 above)
Comment (mc-jones) [reply src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts:20]:
```suggestion
Walk through steps:
```

File: src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts (hunk 7 above)
Comment (mc-jones) [reply src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts:31]: Yeah, good call. Oversight.

File: src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts (hunk 11 above)
Comment (mc-jones) [reply src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts:54]: Yeah, I should have been more explicit about the circumstances that resulting in this. It was when I was unauthenticated, and it asked me directly what my purposed for collecting was. Something like "Do you collect for aesthetic enjoyment, investment opportunities, or a mix of both?". So in that sense it was accurate...but something we would want to guard against.

File: src/05-instruction-experiments/server.ts (hunk 18 above)
Comment (mc-jones) [reply src/05-instruction-experiments/server.ts:97]: I haven't tested this, yet, but wanted to leave the option open, as it seemed reasonable that the model could want to use multiple queries/tools to resolve a prompt with just a bit more complexity. But more testing will be needed to determine if this is a YAGTNI situation. 

We may also not want to test beyond the `me` profile, in which  case we can definitely simplify.

File: src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts (hunk 2 above)
Comment (mc-jones) [reply src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts:42]: I like the PR idea. 

Also wonder if using a commit description to track changes would be a good pattern? This could allow us track smaller changes to individual prompts. But does feel a bit like abusing git descriptions....

File: src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts (hunk 7 above)
Comment (mc-jones) [reply src/05-instruction-experiments/00-prompt-lib/collector-profile-one.ts]: Fixed in [b78848d](https://github.com/artsy/quantum/pull/6/commits/b78848de425d45e5400559c3b6543d9d744bca6e)

**Comment by mc-jones** [comment]
> Sorry for the profusion of comments. 

No need to apologize! I appreciate you taking the time for a the thorough review. Sorry there were so many typos here.

>I imagine it will get tiresome for everyone if we have to keep manually flagging these. 

Yeah, I think that is a good call, as a person whose fingers are often faster than his brain. I've added config in [e5486d0](https://github.com/artsy/quantum/pull/6/commits/e5486d0f0f3542c71907cd20d9edb1d2fde20dd8) to suggest this be install for the repo. Not sure if we want to be more explicit, but seems fine since there are only three of us for now.
